testutil-server/src/main/java/io/spine/testing/server/blackbox/ClientSupplier.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2021, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.testing.server.blackbox;
+
+import io.spine.base.Identifier;
+import io.spine.client.Client;
+import io.spine.core.TenantId;
+import io.spine.server.BoundedContext;
+import io.spine.server.Closeable;
+import io.spine.server.CommandService;
+import io.spine.server.GrpcContainer;
+import io.spine.server.QueryService;
+import io.spine.server.SubscriptionService;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.protobuf.TextFormat.shortDebugString;
+import static io.spine.util.Exceptions.illegalStateWithCauseOf;
+import static java.util.Objects.isNull;
+
+/**
+ * Creates {@code Client} instances for the passed Bounded Context.
+ *
+ * <p>As {@code Client} requires a gRPC server to be connected to, the supplier assembles
+ * and starts a {@link GrpcContainer}. The server and the created clients are working in-process.
+ */
+class ClientSupplier implements Closeable {
```

Comment:
This should be called `Clients` so that it's coherent with the `BlackBoxClients`. Maybe, `ClientFactory`, but then the `BlackBoxClients` should also be renamed.

This type is too complicated to be a "supplier".